src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_ForEachStatement.cs

```diff
@@ -520,12 +519,25 @@ private BoundStatement RewriteStringForEachStatement(BoundForEachStatement node)
                 iterationVarDecl = new BoundExpressionStatement(assignment.Syntax, loweredAssignment);
                 RemovePlaceholderReplacement(deconstruction.TargetPlaceholder);
 
-                iterationVariables = assignment.LeftVariables.SelectAsArray(v => ((BoundLocal)v).LocalSymbol);
+                iterationVariables = GetLocalSymbols(assignment.LeftVariables);
             }
 
             return iterationVarDecl;
         }
 
+        private static ImmutableArray<LocalSymbol> GetLocalSymbols(ImmutableArray<BoundExpression> leftVariables)
+        {
+            var builder = ArrayBuilder<LocalSymbol>.GetInstance();
+            foreach (var variable in leftVariables)
+            {
+                if (variable.Kind == BoundKind.Local)
+                {
+                    builder.Add(((BoundLocal)variable).LocalSymbol);
+                }
+            }
+            return builder.ToImmutableAndFree();
```

Comment:
Yes. See `ForEachIEnumerableDeclarationWithDeconstruct` test which has a foreach with nested declarations and verifies the symbols.

As things stand, the deconstruction variables get flattened at the end of initial binding.

---
In reply to: [88763772](https://github.com/dotnet/roslyn/pull/15192#discussion_r88763772) [](ancestors = 88763772)